bitbots_blackboard/bitbots_blackboard/capsules/costmap_capsule.py

```diff
@@ -0,0 +1,390 @@
+        self.obstacle_cost: float = self._blackboard.node.get_parameter('body.obstacle_cost').value
+
+
+        # Publisher for visualization in RViZ
+        self.costmap_publisher = self._blackboard.node.create_publisher(OccupancyGrid, 'debug/costmap', 1)
+
+        self.base_costmap: Optional[np.ndarray] = None  # generated once in constructor field features
+        self.costmap: Optional[np.ndarray] = None  # updated on the fly based on the base_costmap
+        self.gradient_map: Optional[np.ndarray] = None  # global heading map (static) only dependent on field structure
+
+        # Calculates the base costmap and gradient map based on it
+        self.calc_base_costmap()
+        self.calc_gradients()
+
+
+    ############
+    # Obstacle #
+    ############
+
+    def robot_obstacle_callback(self, msg: RobotArray):
+        """
+        Callback with new obstacles
+        """
+        # Init a new obstacle costmap
+        obstacle_map = np.zeros_like(self.costmap)
+        # Iterate over all obstacles
+        robot: Robot
+        for robot in msg.robots:
+            # Convert position to array index
+            idx_x, idx_y = self.field_2_costmap_coord(robot.bb.center.position.x, robot.bb.center.position.x)
+            # TODO inflate
+            # Draw obstacle with smoothing independent weight on obstacle costmap
+            obstacle_map[idx_x, idx_y] = \
+                self.obstacle_cost * self.obstacle_costmap_smoothing_sigma
+        # Smooth obstacle map
+        obstacle_map = gaussian_filter(obstacle_map, self.obstacle_costmap_smoothing_sigma)
+        # Get pass offsets
+        self.pass_map = self.get_pass_regions()
+        # Merge costmaps
+        self.costmap = self.base_costmap.copy() + obstacle_map - self.pass_map
+        # Publish debug costmap
+        self.costmap_debug_draw()
+
+    def costmap_debug_draw(self):
+        """
+        Publishes the costmap for rviz
+        """
+        # Normalize costmap to match the rviz color scheme in a good way
+        normalized_costmap = (255 - ((self.costmap - np.min(self.costmap)) / (
+                    np.max(self.costmap) - np.min(self.costmap))) * 255 / 2.1).astype(np.int8).T
+        # Build the OccupancyGrid message
+        msg: OccupancyGrid = ros2_numpy.msgify(
+            OccupancyGrid,
+            normalized_costmap,
+            info=MapMetaData(
+                resolution=0.1,
+                origin=Pose(
+                    position=Point(
+                        x=-self.field_length / 2 - self.map_margin,
+                        y=-self.field_width / 2 - self.map_margin,
+                    )
+                )))
+        # Change the frame to allow namespaces
+        msg.header.frame_id = self.map_frame
+        # Publish
+        self.costmap_publisher.publish(msg)
+
+    def get_pass_regions(self) -> np.ndarray:
+        """
+        Draws a costmap for the pass regions
+        """
+        pass_dist = 1.0
```

Comment:
Maybe make these ros parameters

Reply:
Does that mean putting them into the body_behavior.yaml and loading them from self.body_config?